tools/newgtlds.go

```diff
@@ -0,0 +1,254 @@
+	// domains present in the PSL data. Entries present in legacyGTLDs will not be
+	// output by this tool when generating the new gTLD data.
+	legacyGTLDs = map[string]bool{
+		"aero":   true,
+		"asia":   true,
+		"biz":    true,
+		"cat":    true,
+		"com":    true,
+		"coop":   true,
+		"info":   true,
+		"jobs":   true,
+		"mobi":   true,
+		"museum": true,
+		"name":   true,
+		"net":    true,
+		"org":    true,
+		"post":   true,
+		"pro":    true,
+		"tel":    true,
+		"xxx":    true,
+	}
+
+	// pslTemplate is a parsed text/template instance for rendering a list of pslEntry
+	// objects in the format used by the public suffix list.
+	//
+	// It expects the following template data:
+	//   URL - the string URL that the data was fetched from.
+	//   Date - the time.Date that the data was fetched.
+	//   Entries - a list of pslEntry objects.
+	pslTemplate = template.Must(template.New("public-suffix-list-gtlds").Parse(`
+// List of new gTLDs imported from {{ .URL }} on {{ .Date.Format "2006-01-02T15:04:05Z07:00" }}
+// This list is auto-generated, don't edit it manually.
+
+{{- range .Entries }}
+{{ .Comment }}
+{{ printf "%s\n" .ULabel}}
+{{- end }}
+`))
+)
+
+// pslEntry is a struct matching a subset of the gTLD data fields present in
+// each object entry of the "GLTDs" array from ICANN_GTLD_JSON_URL.
+type pslEntry struct {
+	// GTLD contains the gTLD name. For internationalized gTLDs the GTLD field is
+	// expressed in punycode.
+	GTLD string
+	// ULabel contains the unicode representation of the gTLD name. When the gTLD
+	// ULabel in the ICANN gTLD data is empty (e.g for an ASCII gTLD like
+	// '.pizza') the PSL entry will use the GTLD as the ULabel.
+	ULabel string
+	// RegistryOperator holds the name of the registry operator that operates the
+	// gTLD (may be empty).
+	RegistryOperator string
+	// DateOfContractSignature holds the date the gTLD contract was signed (may be empty).
+	DateOfContractSignature string
+	// ContractTerminated indicates whether the contract has been terminated by
+	// ICANN. When rendered by the pslTemplate only entries with
+	// ContractTerminated = false are included.
+	ContractTerminated bool
+}
+
+// normalize will normalize a pslEntry by mutating it in place to trim the
+// string fields of whitespace and by populating the ULabel with the GTLD if
+// the ULabel is empty.
+func (e *pslEntry) normalize() {
+	e.GTLD = strings.TrimSpace(e.GTLD)
+	e.ULabel = strings.TrimSpace(e.ULabel)
+	e.RegistryOperator = strings.TrimSpace(e.RegistryOperator)
+	e.DateOfContractSignature = strings.TrimSpace(e.DateOfContractSignature)
+
+	// If there is no explicit uLabel use the gTLD as the uLabel.
+	if e.ULabel == "" {
+		e.ULabel = e.GTLD
+	}
+}
+
+// Comment generates a comment string for the pslEntry. This string has a `//`
+// prefix and includes:
+//   * the ASCII representation of the gTLD (punycode for internationalized
+//     gTLDs).
+//   * the date of the contract signature (may be empty).
```

Comment:
The description here says "may be empty", but then it's listed as unconditionally included.

The reason this reads a bit weird to me is that, by the same logic, RegistryOperator is also always present, but may be empty as well. That is, it seems the three forms are:

```
<gTLD> :
<gTLD> : <Contract Signature Date>
<gTLD> : <Contract Signature Date> <RegistryOperator>
```

I suppose the only distinction is that, without the RegistryOperator check, we could end up with two trailing spaces instead of one, right?

Reply:
TBH I was trying to exactly match the logic from the previous tool to avoid output diffs without putting much thought into the "why". I think my textual comment probably misrepresented the result.

https://github.com/publicsuffix/list/blob/6f03f42a65d006c8ae657f125f14fb8f9d3337f4/tools/newgtlds#L38-L40

> I suppose the only distinction is that, without the RegistryOperator check, we could end up with two trailing spaces instead of one, right?

Yes, I think that's what the original code was trying to avoid. I'll update my comment to try and make that clearer.

Reply:
I took a crack at writing a better comment in 9ef22c3 - I think it better represents the possible output now.